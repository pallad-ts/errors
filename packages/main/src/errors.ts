import {ErrorDescriptor} from "./ErrorDescriptor";
import {formatCodeFactory, PLACEHOLDER_TOKEN} from "./formatCodeFactory";


const formatCode = formatCodeFactory('E_PALLAD_ERRORS_%c');

// Needed to solve circular reference problem where ErrorDescriptor is not yet defined here
function lazyInitialization<TFactory extends (...args: any[]) => any, TDescriptorFactory extends () => ErrorDescriptor<TFactory, any>>(descriptorFactory: TDescriptorFactory) {
	let descriptor: ErrorDescriptor<any>;
	return (...args: Parameters<TFactory>) => {
		if (!descriptor) {
			descriptor = descriptorFactory();
		}
		return descriptor.create(...args);
	}
}

export const ERRORS = {
	CODE_ALREADY_TAKEN: lazyInitialization(
		() => ErrorDescriptor.useMessageFormatter(formatCode(1), (code: string) => {
			return `Error code "${code}" is already registered`;
		})
	),
	DOMAIN_IS_LOCKED: lazyInitialization(
		() => ErrorDescriptor.useDefaultMessage(formatCode(2), 'Domain is locked and no other error descriptors can be added')
	),
	ERROR_FROZEN: lazyInitialization(
		() => ErrorDescriptor.useDefaultMessage(formatCode(3), 'Cannot assign "code" property to frozen error object')
	),
	MISSING_PLACEHOLDER_IN_CODE_FORMAT: lazyInitialization(
		() => ErrorDescriptor.useMessageFormatter(formatCode(4), (format: string) => {
			return `Code placeholder "${PLACEHOLDER_TOKEN}" is missing in format "${format}"`;
		})
	),
	INVALID_ERROR_DESCRIPTOR_UNDER_KEY: lazyInitialization(
		() => ErrorDescriptor.useMessageFormatter(formatCode(5), (key: string) => {
			return `Value under key "${key}" is not valid ErrorDescriptor`;
		})
	)
};

