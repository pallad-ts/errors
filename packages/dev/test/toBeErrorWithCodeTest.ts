import {ErrorDescriptor} from "@pallad/errors";

describe('toBeErrorWithCode', () => {
	const descriptor = ErrorDescriptor.useDefaultMessage('C_1', 'Some default message');
	const descriptorSecond = ErrorDescriptor.useDefaultMessage('C_2', 'Some default message 2');

	it('passes if error with proper code is provided', () => {
		expect(descriptor.create())
			.toBeErrorWithCode(descriptor);
	});

	it('passes if error with proper code is provided (as code)', () => {
		expect(descriptor.create())
			.toBeErrorWithCode(descriptor.code);
	});

	it('passes if error code matches one of provided codes', () => {
		expect(descriptor.create())
			.toBeErrorWithCode(descriptor, descriptorSecond);
		expect(descriptorSecond.create())
			.toBeErrorWithCode(descriptor, descriptorSecond);
	})

	it('passes if error (as promise) with proper code is provided', () => {
		return expect(Promise.resolve(descriptor.create()))
			.resolves
			.toBeErrorWithCode(descriptor);
	});

	it('fails if error with different code is provided', () => {
		expect(() => {
			expect(descriptor.create())
				.toBeErrorWithCode(descriptorSecond);
		})
			.toThrowErrorMatchingSnapshot();
	})
});
