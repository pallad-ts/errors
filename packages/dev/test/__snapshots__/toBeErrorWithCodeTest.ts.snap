// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeErrorWithCode fails if error with different code is provided 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeErrorWithCode([22m[32mexpected[39m[2m)[22m

Expected error with code: C_2
Received: [31m[Error: Some default message][39m with code: C_1"
`;
